tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,251 @@
+
+        if let Poll::Ready(result) = this.inner.poll(cx) {
+            return Poll::Ready(result.map_err(Into::into));
+        }
+
+        if let OptionPinProj::Some(sleep) = this.sleep.project() {
+            futures_util::ready!(sleep.poll(cx));
+            return Poll::Ready(Err(TimeoutExpired.into()));
+        }
+
+        Poll::Pending
+    }
+}
+
+#[pin_project(project = OptionPinProj)]
+enum OptionPin<T> {
+    Some(#[pin] T),
+    None,
+}
+
+const GRPC_TIMEOUT_HEADER: &str = "grpc-timeout";
+
+const SECONDS_IN_HOUR: u64 = 60 * 60;
+const SECONDS_IN_MINUTE: u64 = 60;
+
+/// Tries to parse the `grpc-timeout` header if it is present. If we fail to parse, returns
+/// the value we attempted to parse.
+///
+/// Follows the [gRPC over HTTP2 spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).
+fn try_parse_grpc_timeout(
+    headers: &HeaderMap<HeaderValue>,
+) -> Result<Option<Duration>, &HeaderValue> {
+    match headers.get(GRPC_TIMEOUT_HEADER) {
+        Some(val) => {
+            let (timeout_value, timeout_unit) = val
+                .to_str()
+                .map_err(|_| val)
+                .and_then(|s| if s.is_empty() { Err(val) } else { Ok(s) })?
+                // `HeaderValue::to_str` only returns `Ok` if the header contains ASCII so this
+                // `split_at` will never panic from trying to split in the middle of a character.
+                // See https://docs.rs/http/0.2.4/http/header/struct.HeaderValue.html#method.to_str
+                //
+                // `len - 1` also wont panic since we just checked `s.is_empty`.
+                .split_at(val.len() - 1);
+
+            // gRPC spec specifies `TimeoutValue` will be at most 8 digits
+            // Caping this at 8 digits also prevents integer overflow from ever occurring
+            if timeout_value.len() > 8 {
+                return Err(val);
+            }
+
+            let timeout_value: u64 = timeout_value.parse().map_err(|_| val)?;
+
+            let duration = match timeout_unit {
+                // Hours
+                "H" => Duration::from_secs(timeout_value * SECONDS_IN_HOUR),
+                // Minutes
+                "M" => Duration::from_secs(timeout_value * SECONDS_IN_MINUTE),
+                // Seconds
+                "S" => Duration::from_secs(timeout_value),
+                // Milliseconds
+                "m" => Duration::from_millis(timeout_value),
+                // Microseconds
+                "u" => Duration::from_micros(timeout_value),
+                // Nanoseconds
+                "n" => Duration::from_nanos(timeout_value),
+                _ => return Err(val),
+            };
+
+            Ok(Some(duration))
+        }
+        None => Ok(None),
+    }
+}
+
+// Note: The wrapped Duration should only be used for logging purposes. It is **not** the
+// actual duration that elapsed, resulting in a timeout, instead it is a close approximation
+#[derive(Debug)]
+struct TimeoutExpired;
```

Comment:
```suggestion
struct TimeoutExpired(());
```

We might want to expose this so people can downcast to it?

Reply:
We also probably should actually enumerate all the errors.....lol

Reply:
Out of curiosity, whats the difference between `struct TimeoutExpired` and struct `TimeoutExpired(())`?

Reply:
It allows someone to never construct the type and thus would allow us to add new stuff to it without breaking the type. Because the internal `0` indexed tuple item is private. So just good tactic to ensure you have flexibility in the future.

Reply:
Good idea!